prow/gerrit/client/client.go

```diff
@@ -159,47 +163,79 @@ func NewClient(instances map[string][]string) (*Client, error) {
 	return c, nil
 }
 
-func auth(c *Client, cookiefilePath string) {
-	logrus.Info("Starting auth loop...")
-	var previousToken string
-	wait := time.Minute
-	for {
-		raw, err := ioutil.ReadFile(cookiefilePath)
-		if err != nil {
-			logrus.WithError(err).Error("Failed to read auth cookie")
-		}
-		fields := strings.Fields(string(raw))
-		token := fields[len(fields)-1]
+func (c *Client) authenticateOnce(previousToken string) string {
+	c.lock.RLock()
+	auth := c.authentication
+	c.lock.RUnlock()
 
-		if token == previousToken {
-			time.Sleep(wait)
-			continue
-		}
+	current, err := auth()
+	if err != nil {
+		logrus.WithError(err).Error("Failed to read gerrit auth token")
+	}
+
+	if current == previousToken {
+		return current
+	}
 
-		logrus.Info("New token, updating handlers...")
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	logrus.Info("New gerrit token, updating handlers...")
 
-		// update auth token for each instance
-		for instance, handler := range c.handlers {
-			handler.authService.SetCookieAuth("o", token)
+	// update auth token for each instance
+	for instance, handler := range c.handlers {
+		handler.authService.SetCookieAuth("o", current)
 
-			self, _, err := handler.accountService.GetAccount("self")
-			if err != nil {
-				logrus.WithError(err).Error("Failed to auth with token")
-				continue
-			}
-			logrus.Infof("Authentication to %s successful, Username: %s", handler.instance, self.Name)
-			c.accounts[instance] = self
+		self, _, err := handler.accountService.GetAccount("self")
+		if err != nil {
+			logrus.WithField("instance", instance).WithError(err).Error("Failed to auth with token")
+			continue
 		}
-		previousToken = token
-		time.Sleep(wait)
+		logrus.Infof("Authentication to %s successful, Username: %s", handler.instance, self.Name)
+		c.accounts[instance] = self
 	}
+	return current
 }
 
-// Start will authenticate the client with gerrit periodically
-// Start must be called before user calls any client functions.
-func (c *Client) Start(cookiefilePath string) {
-	if cookiefilePath != "" {
-		go auth(c, cookiefilePath)
+// Authenticate client calls using the specified file.
+// Periodically re-reads the file to check for an updated value.
+func (c *Client) Authenticate(cookiefilePath, tokenPath string) {
+	var was, auth func() (string, error)
+	switch {
+	case cookiefilePath != "":
+		auth = func() (string, error) {
+			// TODO(fejta): listen for changes
+			raw, err := ioutil.ReadFile(cookiefilePath)
+			if err != nil {
+				return "", fmt.Errorf("read cookie: %w", err)
+			}
+			fields := strings.Fields(string(raw))
+			token := fields[len(fields)-1]
+			return token, nil
+		}
+	case tokenPath != "":
```

Comment:
nit: do you want the override to be the first case?

Reply:
Good point, looks like there's more stuff to futz with so will address in a followup